src/control/cmd/dmg/system.go

```diff
@@ -161,26 +163,55 @@ func displaySystemQuerySingle(log logging.Logger, members system.Members) error
 	return nil
 }
 
+// rankListCmd enables rank or host list to be supplied with command to filter
+// which ranks are operated upon.
+type rankListCmd struct {
+	RankList string `long:"ranks" short:"r" description:"Comma separated ranges or individual system ranks to operate on"`
+	HostList string `long:"rank-hosts" short:"h" description:"Hostlist representing hosts whose managed ranks are to be operated on"`
+}
+
+// validateRanksHosts validates rank and host lists have correct format.
+func (cmd *rankListCmd) validateRanksHosts() error {
+	hasRanks := len(cmd.RankList) > 0
+	hasHosts := len(cmd.HostList) > 0
+
+	switch {
+	case hasRanks && hasHosts:
+		return errors.New("--ranks and --rank-hosts options cannot be set together")
+	case hasRanks:
+		if _, err := system.ParseRanks(cmd.RankList); err != nil {
+			return errors.Wrap(err, "ranks list")
+		}
+	case hasHosts:
+		if _, err := hostlist.CreateSet(cmd.HostList); err != nil {
+			return errors.Wrap(err, "rank-hosts list")
+		}
+	}
+
+	return nil
+}
+
 // systemQueryCmd is the struct representing the command to query system status.
 type systemQueryCmd struct {
 	logCmd
 	ctlInvokerCmd
 	jsonOutputCmd
-	Verbose bool   `long:"verbose" short:"v" description:"Display more member details"`
-	Ranks   string `long:"ranks" short:"r" description:"Comma separated ranges or individual system ranks to query"`
+	rankListCmd
+	Verbose bool `long:"verbose" short:"v" description:"Display more member details"`
 }
 
 // Execute is run when systemQueryCmd activates
 func (cmd *systemQueryCmd) Execute(_ []string) error {
-	ranks, err := system.ParseRanks(cmd.Ranks)
-	if err != nil {
-		return errors.Wrap(err, "parsing rank list")
+	if err := cmd.validateRanksHosts(); err != nil {
+		return err
 	}
 
-	ctx := context.Background()
+	req := new(control.SystemQueryReq)
+	if err := convert.Types(cmd, req); err != nil {
```

Comment:
Not sure how I feel about this... I think it would be clearer to be more explicit about the request fields and which inputs are being taken from the cmd struct. The `convert.Types()` helper was intended more for switching between protobuf and native structs.